### PR TITLE
Add `fstatat()` availibility autodetection/configuration and replace with `stat()` if necessary

### PR DESCRIPTION
**Pull request by Ionic**
Fixes: DanBloomberg/leptonica#327

### CODE REVIEW

**Comment by DanBloomberg** [comment]
Thank you for fixing the problem introduced in January with fstatat().

**Comment by Ionic** [comment]
You're welcome.

There is one caveat, though: if you were to rely on `fstatat()`-specific behavior (like passing non-zero flags or relying on it treating a relative path relative to the file descriptor's current location instead of current to cwd), then `stat()` is not a proper replacement for `fstatat()`.

To the best of my knowledge, as far as looking through your code went, I couldn't determine anything relying on `fstatat()`-specific behavior, but if the code is changed developers should definitely keep that in mind.

File: src/sarray1.c [hunk 1]
```diff
@@ -1871,9 +1871,14 @@ struct stat     st;
     safiles = sarrayCreate(0);
     dfd = dirfd(pdir);
     while ((pdirentry = readdir(pdir))) {
+#if HAVE_FSTATAT
+      stat_ret = fstatat(dfd, pdirentry->d_name, &st, 0);
+#else
+      stat_ret = stat(pdirentry->d_name, &st);
```
Comment (ip1981) [inline src/sarray1.c]: This depends on the current directory.

**Comment by ip1981** [comment]
I could not care less about old non-free OSes, but if this is a problem, removing this code is the next best option :) Either unconditionally or when `fstatat` is absent.

**Comment by DanBloomberg** [comment]
OK, help me out here.  When you say "removing this code is the next best option", what code are you referring to?

 I really have 3 questions:
* does anything need to be changed, and if so, why?
* do we have consensus on that?

  -- Dan

**Comment by ip1981** [comment]
I refer to the code that detects directories. How useful is it?

Existing code should work just fine with `fstatat`, but not with `stat` .
This PR addresses compilation issues only.

I remember I fixed both compilation and runtime issues (reiserfs AFAIK does not set  `d_type` on Linux, no `d_type` on Solaris).

So again, how critical is this feature (skipping on directories)?

**Comment by DanBloomberg** [comment]
It may not be very important for most applications.  But if the information is available, why not use it?

I did some experiments.  As you say, it all works with fstatat.
But with stat, it finds "." and ".." as directories but not subdirectories.

Why not just put in a comment to that effect?

**Comment by DanBloomberg** [comment]
I did that & pushed to the master.   Are we ready to close this issue?

**Comment by Ionic** [comment]
`stat()` and `fstatat()` behave exactly the same if passed an absolute path (assuming the `flags` are `0`). Using `realpath` on `realdir` and prepending this to additional directory entries when calling `stat()` (only) would work around any dependence upon `CWD`.

**Comment by DanBloomberg** [comment]
When I do this (calling stat, not fstatat):

    char *realp = realpath("lion", NULL);
    Sarray *sa = getFilenamesInDirectory(realp);

I still get the subdirectories of "lion" added into sa.
What flags are supposed to be 0?

**Comment by Ionic** [comment]
Careful with passing `NULL` to `realpath()`. This invokes implementation-defined behavior [in older versions of SUS/POSIX](http://pubs.opengroup.org/onlinepubs/009695399/functions/realpath.html) and some implementions crash in such a case. [Newer versions allow this, making `realpath()` malloc a buffer by itself](http://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html), but of course that is not widely supported yet. It's better to always give it a proper `PATH_MAX + 1`-length buffer. (That, btw., is in line with the lot of `str*` functions that expect proper strings and often crash if passing `NULL`.)

`When I do this (calling stat, not fstatat):`

You'll also have to prepend the absolute base path to the filename when calling stat, i.e.,

```C
#if HAVE_FSTATAT
        stat_ret = fstatat(dfd, pdirentry->d_name, &st, 0);
#else
        size_t tmp_size = strlen(realdir) + strlen(pdirentry->d_name) + 1; /* Additional slash */
        char *stat_path = malloc(tmp_size + 1); /* Terminating NULL */
        snprintf(stat_path, tmp_size + 1, "%s/%s", realdir, pdirentry->d_name);
        stat_ret = stat(stat_path, &st);
        free(stat_path);
        stat_path = NULL;
#endif
```

**Comment by DanBloomberg** [comment]
Thanks!  That worked well.  I think this resolves the issue, as now stat() and fstatat() both handle directories properly.

**Comment by Ionic** [comment]
Yeah, but `realpath()` should also be executed unconditionally in the `stat()` case, since the function shouldn't rely upon being provided an absolute path... at least in theory.

**Comment by ip1981** [comment]
Oh, no :) multiple allocations and deallocations without checking for NULL. Is the problem really worth it?

**Comment by Ionic** [comment]
That was just meant as a quick way for Dan to demonstrate what I was referring to.

Naturally in the actual code the return values of `malloc()`, `snprintf()` and `realpath()` (which should be inserted into that function in the `stat()`-branch as well, instead of assuming that an actual absolute path is always passed in) should all be checked and the function error'd out properly.

Whether or not that's worth it... I guess it depends. Changing behavior if `fstatat()` is not available sounds worse to me than a compat shim. While `fstatat()` was commonly available on Linux-based platforms, `*BSD` ones only adopted out ~10 years ago. That's a long time span, but why break stuff if it's so easy to provide an alternative?

**Comment by DanBloomberg** [comment]
Thank you again for all the help.  I've now made changes so that it gets the same (correct) results with either stat() or fstatat().

It doesn't need to check for the return of realpath() because we're not allocating memory there -- just writing to a buffer on the stack.

Are we ready to close this issue?

**Comment by Ionic** [comment]
Should be `PATH_MAX + 1` (since the macro does not include a NULL terminator).

Like @ip1981 suggested, it's also worthwhile to
```C
    if (!stat_path)
        return (SARRAY *)ERROR_PTR("unable to allocate memory for stat_path", procName, NULL);
```

after the `LEPT_CALLOC` call for `stat_path`.


Plus, check `snprintf`'s return code:
```C
    size_t snprintf_ret = snprintf(stat_path, size - 1, "%s/%s", realdir, pdirentry->d_name);
    if ((snprintf_ret < 0) || (snprintf_ret >= size)
        return (SARRAY *)ERROR_PTR("unable to construct absolute path", procName, NULL);
```

There is also a problem with the size expression provided there: you add `+ 2` for both the slash and `NULL` terminator. In that case, you can just alloc `size` (not `size + 1` bytes), but will have to pass `size - 1` to `snprintf`, since the `n` denotes the number of characters that are to written minus the always written NULL byte (yeah, `snprintf` is a bit hairy). I've corrected this in my previous code example for the `snprintf` call - the return value check is also correct, because it denotes the bytes that would have been written if `n` was large enough, excluding the terminating NULL byte (hence, if it's returning a number of `size` or bigger you know that the output was truncated).

Otherwise I'm content.

**Comment by Ionic** [comment]
Should have been `ptrdiff_t` instead of `size_t` of course, since the return code can be negative.

**Comment by DanBloomberg** [comment]
Thank you again for reviewing the code. It is finicky to get this done well!

I took some of your advice, and did it a bit differently.
I checked the size against PATH_MAX and skipped with an error message if
too large.
Otherwise, we know the calloc will succeed (it's small), so no need to
check that.
The snprintf() needs size, not size - 1, to get all the characters, and
again no need to check the return since we know the buffer is big enough
(it is PATH_MAX + 1, as you suggested).

**Comment by Ionic** [comment]
> I checked the size against PATH_MAX and skipped with an error message if too large.

Okay, sane.

> Otherwise, we know the calloc will succeed (it's small), so no need to check that.

Well... strictly speaking, memory exhaustion (i.e., big requests) is not the only cause of `malloc` failures; if the program's memory space is heavily fragmented, allocating even smaller requests may fail. I typically check all allocations, no matter how small, but it's your call of course. :)

> The snprintf() needs size, not size - 1, to get all the characters

Right, there was that... it's the total buffer size including the terminating `NULL` character, sorry. I typically check `snprintf` return codes as well, even when I know that the buffer is big enough, but that might be overly defensive.

Anyway, I guess I'm fine with all of that. Unless Igor has other comments, we're probably done here.

**Comment by ip1981** [comment]
There is gnulib which provides drop-in replacement for many functions, including for `fstatat` :)

Anyway, I think this feature is not worth that much of [sloppy] code. As I mentioned in my original PR I'd delete it.

Depending on the use case, there might be another quite portable option - `ftw`, http://pubs.opengroup.org/onlinepubs/007904975/functions/ftw.html

**Comment by DanBloomberg** [comment]
I agree that a lot of time has been spent on this, but it has been educational for me, and the code is now working correctly in all cases.

From the beginning, I thought of leptonica as having two main goals: (1) the didactic one of showing how to make useful image processing and analysis code (with decent C programming), and (2) providing a library that can be used for both development (because of simplicity) and production (because of efficiency).  Getting this function to work properly seems to satisfy both goals, and I appreciate the effort you both made to get it to that point.

**Pull request merged**